handler/src/main/java/io/netty/handler/ssl/SslHandler.java

```diff
@@ -641,7 +646,14 @@ private void wrapNonAppData(ChannelHandlerContext ctx, boolean inUnwrap) throws
                 }
             }
         } catch (SSLException e) {
-            setHandshakeFailure(ctx, e);
+            try {
+                setHandshakeFailure(ctx, e);
+            } finally {
+                // We may have written some parts of data before an exception was thrown so ensure we always flush.
+                // See https://github.com/netty/netty/issues/3900#issuecomment-172481830
+                mayFlush(ctx);
```

Comment:
@normanmaurer - thanks for the clarification. i took a deeper look and the current approach seems reasonable. gets interesting when wrap and unwrap can call each other.